codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
         return notNull
     }
 
+    private fun mapFunctionLogic(
+        expression: JmespathExpression,
+        arg: VisitedExpression,
+        resultName: String,
+        operation: String,
+        stringOrNumberCheck: Boolean = true,
+    ): String {
+        val argName = arg.identifier
+        val result = bestTempVarName(resultName)
+        writer.withBlock("val $result = $argName?.$operation{", "}") {
+            val expressionValue = addTempVar("expression", subfieldCodegen((expression as ExpressionTypeExpression).expression as FieldExpression, "it"))
+            if (stringOrNumberCheck) {
+                write("if ($expressionValue as Any !is Number && $expressionValue as Any !is String) throw Exception(\"Result of applying expression should be string or number\")")
+            }
```

Comment:
Same comment about `invertedArgs` here, why is this needed for `map` only?

Reply:
The [map](https://jmespath.org/specification.html#map) expression can result in any type but for the rest ([max_by](https://jmespath.org/specification.html#max-by), `min_by`, etc..) the expression should result in a string or a number.